File: backend/pkg/theme/theme.go
package theme

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
)

const SettingsKey = "theme"

var hexColorRE = regexp.MustCompile(`^#[0-9A-Fa-f]{6}$`)

// Palette is one light or dark color set for the whole platform.
type Palette struct {
	Primary             string `json:"primary"`
	PrimaryHover        string `json:"primary_hover"`
	PrimaryPressed      string `json:"primary_pressed"`
	PrimarySurface      string `json:"primary_surface"`
	Background          string `json:"background"`
	BackgroundSecondary string `json:"background_secondary"`
	BackgroundTertiary  string `json:"background_tertiary"`
	TextPrimary         string `json:"text_primary"`
	TextSecondary       string `json:"text_secondary"`
	TextTertiary        string `json:"text_tertiary"`
	Border              string `json:"border"`
	BorderStrong        string `json:"border_strong"`
	Error               string `json:"error"`
	Success             string `json:"success"`
	Warning             string `json:"warning"`
	BubbleOwn           string `json:"bubble_own"`
	BubbleOthers        string `json:"bubble_others"`
}

// Settings is the platform-wide theme stored in platform_settings.
type Settings struct {
	Light Palette `json:"light"`
	Dark  Palette `json:"dark"`
}

// Default returns the Xilo design-system palette (ui-ux-spec §2).
func Default() Settings {
	return Settings{
		Light: Palette{
			Primary:             "#1D9BF0",
			PrimaryHover:        "#1A8CD8",
			PrimaryPressed:      "#1A7BC5",
			PrimarySurface:      "#E8F5FE",
			Background:          "#FFFFFF",
			BackgroundSecondary: "#F7F9FA",
			BackgroundTertiary:  "#EFF3F4",
			TextPrimary:         "#0F1419",
			TextSecondary:       "#536471",
			TextTertiary:        "#8295A3",
			Border:              "#EFF3F4",
			BorderStrong:        "#CFD9DE",
			Error:               "#F4212E",
			Success:             "#00BA7C",
			Warning:             "#FFAD1F",
			BubbleOwn:           "#E8F5FE",
			BubbleOthers:        "#F7F9FA",
		},
		Dark: Palette{
			Primary:             "#1D9BF0",
			PrimaryHover:        "#4DB8F5",
			PrimaryPressed:      "#6BC9F7",
			PrimarySurface:      "#1A2A3A",
			Background:          "#15202B",
			BackgroundSecondary: "#192734",
			BackgroundTertiary:  "#22303C",
			TextPrimary:         "#E7E9EA",
			TextSecondary:       "#71767B",
			TextTertiary:        "#536471",
			Border:              "#38444D",
			BorderStrong:        "#4A5A66",
			Error:               "#F4212E",
			Success:             "#00BA7C",
			Warning:             "#FFAD1F",
			BubbleOwn:           "#1E3A5F",
			BubbleOthers:        "#2C2C2E",
		},
	}
}

// Validate ensures every color field is a #RRGGBB hex value.
func Validate(s Settings) error {
	if err := validatePalette("light", s.Light); err != nil {
		return err
	}
	return validatePalette("dark", s.Dark)
}

func validatePalette(mode string, p Palette) error {
	fields := map[string]string{
		"primary":              p.Primary,
		"primary_hover":        p.PrimaryHover,
		"primary_pressed":      p.PrimaryPressed,
		"primary_surface":      p.PrimarySurface,
		"background":           p.Background,
		"background_secondary": p.BackgroundSecondary,
		"background_tertiary":  p.BackgroundTertiary,
		"text_primary":         p.TextPrimary,
		"text_secondary":       p.TextSecondary,
		"text_tertiary":        p.TextTertiary,
		"border":               p.Border,
		"border_strong":        p.BorderStrong,
		"error":                p.Error,
		"success":              p.Success,
		"warning":              p.Warning,
		"bubble_own":           p.BubbleOwn,
		"bubble_others":        p.BubbleOthers,
	}
	for name, value := range fields {
		if !IsHexColor(value) {
			return fmt.Errorf("invalid %s.%s color: %q (expected #RRGGBB)", mode, name, value)
		}
	}
	return nil
}

// IsHexColor reports whether s is a 6-digit hex color with leading #.
func IsHexColor(s string) bool {
	return hexColorRE.MatchString(strings.TrimSpace(s))
}

// Normalize uppercases hex digits for stable storage.
func Normalize(s Settings) Settings {
	return Settings{
		Light: normalizePalette(s.Light),
		Dark:  normalizePalette(s.Dark),
	}
}

func normalizePalette(p Palette) Palette {
	return Palette{
		Primary:             strings.ToUpper(strings.TrimSpace(p.Primary)),
		PrimaryHover:        strings.ToUpper(strings.TrimSpace(p.PrimaryHover)),
		PrimaryPressed:      strings.ToUpper(strings.TrimSpace(p.PrimaryPressed)),
		PrimarySurface:      strings.ToUpper(strings.TrimSpace(p.PrimarySurface)),
		Background:          strings.ToUpper(strings.TrimSpace(p.Background)),
		BackgroundSecondary: strings.ToUpper(strings.TrimSpace(p.BackgroundSecondary)),
		BackgroundTertiary:  strings.ToUpper(strings.TrimSpace(p.BackgroundTertiary)),
		TextPrimary:         strings.ToUpper(strings.TrimSpace(p.TextPrimary)),
		TextSecondary:       strings.ToUpper(strings.TrimSpace(p.TextSecondary)),
		TextTertiary:        strings.ToUpper(strings.TrimSpace(p.TextTertiary)),
		Border:              strings.ToUpper(strings.TrimSpace(p.Border)),
		BorderStrong:        strings.ToUpper(strings.TrimSpace(p.BorderStrong)),
		Error:               strings.ToUpper(strings.TrimSpace(p.Error)),
		Success:             strings.ToUpper(strings.TrimSpace(p.Success)),
		Warning:             strings.ToUpper(strings.TrimSpace(p.Warning)),
		BubbleOwn:           strings.ToUpper(strings.TrimSpace(p.BubbleOwn)),
		BubbleOthers:        strings.ToUpper(strings.TrimSpace(p.BubbleOthers)),
	}
}

// Merge fills empty fields on patch from base (partial admin updates).
func Merge(base, patch Settings) Settings {
	out := base
	out.Light = mergePalette(base.Light, patch.Light)
	out.Dark = mergePalette(base.Dark, patch.Dark)
	return out
}

func mergePalette(base, patch Palette) Palette {
	out := base
	set := func(dst *string, src string) {
		if strings.TrimSpace(src) != "" {
			*dst = src
		}
	}
	set(&out.Primary, patch.Primary)
	set(&out.PrimaryHover, patch.PrimaryHover)
	set(&out.PrimaryPressed, patch.PrimaryPressed)
	set(&out.PrimarySurface, patch.PrimarySurface)
	set(&out.Background, patch.Background)
	set(&out.BackgroundSecondary, patch.BackgroundSecondary)
	set(&out.BackgroundTertiary, patch.BackgroundTertiary)
	set(&out.TextPrimary, patch.TextPrimary)
	set(&out.TextSecondary, patch.TextSecondary)
	set(&out.TextTertiary, patch.TextTertiary)
	set(&out.Border, patch.Border)
	set(&out.BorderStrong, patch.BorderStrong)
	set(&out.Error, patch.Error)
	set(&out.Success, patch.Success)
	set(&out.Warning, patch.Warning)
	set(&out.BubbleOwn, patch.BubbleOwn)
	set(&out.BubbleOthers, patch.BubbleOthers)
	return out
}

// Parse unmarshals theme JSON or returns defaults on empty/invalid structure.
func Parse(raw json.RawMessage) (Settings, error) {
	if len(raw) == 0 {
		return Default(), nil
	}
	var s Settings
	if err := json.Unmarshal(raw, &s); err != nil {
		return Settings{}, err
	}
	merged := Merge(Default(), s)
	if err := Validate(merged); err != nil {
		return Settings{}, err
	}
	return Normalize(merged), nil
}

File: web/src/lib/theme.ts
/** Platform theme palette — matches backend pkg/theme and ui-ux-spec §2. */

export interface ThemePalette {
  primary: string;
  primary_hover: string;
  primary_pressed: string;
  primary_surface: string;
  background: string;
  background_secondary: string;
  background_tertiary: string;
  text_primary: string;
  text_secondary: string;
  text_tertiary: string;
  border: string;
  border_strong: string;
  error: string;
  success: string;
  warning: string;
  bubble_own: string;
  bubble_others: string;
  chat_bubble_own: string;
  chat_bubble_others: string;
}

export interface PlatformTheme {
  light: ThemePalette;
  dark: ThemePalette;
}

export const DEFAULT_THEME: PlatformTheme = {
  light: {
    primary: "#1D9BF0",
    primary_hover: "#1A8CD8",
    primary_pressed: "#1A7BC5",
    primary_surface: "#E8F5FE",
    background: "#FFFFFF",
    background_secondary: "#F7F9FA",
    background_tertiary: "#EFF3F4",
    text_primary: "#0F1419",
    text_secondary: "#536471",
    text_tertiary: "#8295A3",
    border: "#EFF3F4",
    border_strong: "#CFD9DE",
    error: "#F4212E",
    success: "#00BA7C",
    warning: "#FFAD1F",
    bubble_own: "#E8F5FE",
    bubble_others: "#F7F9FA",
    chat_bubble_own: "#E8F5FE",
    chat_bubble_others: "#F0F2F5",
  },
  dark: {
    primary: "#1D9BF0",
    primary_hover: "#4DB8F5",
    primary_pressed: "#6BC9F7",
    primary_surface: "#1A2A3A",
    background: "#15202B",
    background_secondary: "#192734",
    background_tertiary: "#22303C",
    text_primary: "#E7E9EA",
    text_secondary: "#71767B",
    text_tertiary: "#536471",
    border: "#38444D",
    border_strong: "#4A5A66",
    error: "#F4212E",
    success: "#00BA7C",
    warning: "#FFAD1F",
    bubble_own: "#1E3A5F",
    bubble_others: "#2C2C2E",
    chat_bubble_own: "#1E3A5F",
    chat_bubble_others: "#2C2C2E",
  },
};

const HEX_RE = /^#[0-9A-Fa-f]{6}$/;

export function isHexColor(value: string | null | undefined): boolean {
  if (typeof value !== "string") return false;
  return HEX_RE.test(value.trim());
}

/** Fill missing palette keys from defaults (older API payloads omit newer fields). */
export function mergeTheme(partial?: Partial<PlatformTheme> | null): PlatformTheme {
  return {
    light: { ...DEFAULT_THEME.light, ...(partial?.light ?? {}) },
    dark: { ...DEFAULT_THEME.dark, ...(partial?.dark ?? {}) },
  };
}

/** Map palette fields onto --xilo-* CSS variables consumed by globals.css. */
export function applyThemeToDocument(theme: PlatformTheme): void {
  if (typeof document === "undefined") return;
  const root = document.documentElement;
  setPaletteVars(root, "xilo", theme.light);
  setPaletteVars(root, "xilo-dark", theme.dark);
}

function setPaletteVars(root: HTMLElement, prefix: string, p: ThemePalette): void {
  root.style.setProperty(`--${prefix}-background`, p.background);
  root.style.setProperty(`--${prefix}-foreground`, p.text_primary);
  root.style.setProperty(`--${prefix}-card`, p.background);
  root.style.setProperty(`--${prefix}-card-foreground`, p.text_primary);
  root.style.setProperty(`--${prefix}-popover`, p.background);
  root.style.setProperty(`--${prefix}-popover-foreground`, p.text_primary);
  root.style.setProperty(`--${prefix}-primary`, p.primary);
  root.style.setProperty(`--${prefix}-primary-foreground`, "#FFFFFF");
  root.style.setProperty(`--${prefix}-secondary`, p.background_secondary);
  root.style.setProperty(`--${prefix}-secondary-foreground`, p.text_primary);
  root.style.setProperty(`--${prefix}-muted`, p.background_tertiary);
  root.style.setProperty(`--${prefix}-muted-foreground`, p.text_secondary);
  root.style.setProperty(`--${prefix}-accent`, p.primary_surface);
  root.style.setProperty(`--${prefix}-accent-foreground`, p.text_primary);
  root.style.setProperty(`--${prefix}-destructive`, p.error);
  root.style.setProperty(`--${prefix}-border`, p.border);
  root.style.setProperty(`--${prefix}-input`, p.border);
  root.style.setProperty(`--${prefix}-ring`, p.primary);
  root.style.setProperty(`--${prefix}-bubble-own`, p.bubble_own);
  root.style.setProperty(`--${prefix}-bubble-others`, p.bubble_others);
  root.style.setProperty(`--${prefix}-chat-bubble-own`, p.chat_bubble_own);
  root.style.setProperty(`--${prefix}-chat-bubble-others`, p.chat_bubble_others);
}

export const THEME_FIELD_LABELS: { key: keyof ThemePalette; label: string }[] = [
  { key: "primary", label: "رنگ اصلی" },
  { key: "primary_hover", label: "هاور اصلی" },
  { key: "primary_pressed", label: "فشرده اصلی" },
  { key: "primary_surface", label: "سطح اصلی" },
  { key: "background", label: "پس‌زمینه" },
  { key: "background_secondary", label: "پس‌زمینه ثانویه" },
  { key: "background_tertiary", label: "سطح / چیپ‌ها" },
  { key: "text_primary", label: "متن اصلی" },
  { key: "text_secondary", label: "متن ثانویه" },
  { key: "text_tertiary", label: "متن کم‌رنگ" },
  { key: "border", label: "حاشیه" },
  { key: "border_strong", label: "حاشیه قوی" },
  { key: "error", label: "خطا" },
  { key: "success", label: "موفقیت" },
  { key: "warning", label: "هشدار" },
  { key: "bubble_own", label: "حباب خودم" },
  { key: "bubble_others", label: "حباب دیگران" },
  { key: "chat_bubble_own", label: "حباب چت خودم" },
  { key: "chat_bubble_others", label: "حباب چت دیگران" },
];
